data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeCoercionService.java

```diff
@@ -24,10 +24,44 @@ public ParseTreeCoercionService(final Map<Class<? extends Serializable>, Functio
         this.literalTypeConversions = literalTypeConversions;
     }
 
+    public Object applyFunction(final String functionName, final String[] args, Event event) {
+        // Supports length(String) and length(JsonPointer)
+        // For example., length("abcd"), length("/keyName")
+        if (functionName.equals("length")) {
```

Comment:
We should take a factory-based approach similar to what we did for operators. See the `OperatorProvider` class and how it is used.

At a high-level, The `OperatorProvider` takes a `List<Operator<?>>` as input. Spring will inject this list. It builds it up from all the actual `Operator` implementations (see `AndOperator` as one example). Because they are `@Named`, they are injected into the application context and thus become part of the `List<Operator<?>>`.

I think the interface might look something like the following:

```
interface ExpressionFunction<T> {
    int getNumberOfArguments();
    String getFunctionName();
    T evaluate(final Object ... args);
}
```

Reply:
Not sure if this suggestion works for functions unless we plan to put function names in the grammar. Is that what you are suggesting?

Reply:
We discussed offline. No, this will not be a change to the grammar. This will use the Spring application context to get the `List<ExpressionFunction>`.

A class like this should work:

```
@Named
public class ExpressionFunctionProvider {

    private final Map<String, ExpressionFunction> functionsMap;

    @Inject
    public ExpressionFunctionProvider(List<ExpressionFunction> expressionFunctions) {
        functionsMap = expressionFunctions.stream().collect(Collectors.toMap(e -> e.getFunctionName(), e -> e));
    }
    
    public Object provideFunction(String functionName, String[] args, Event event) {
        return functionsMap.get(functionName);
    }
}
```